SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -1849,52 +1881,100 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
 			return;
 		}
-		else
+		else // bUseEntityInteractionSemantics && (bNeedSender || bIsReliable)
 		{
-			AActor* SenderActor = nullptr;
+			// NOTE : the (!bHasSenderAvailable) branch is only there to allow migration to take place
+			// When no sender is available, the RPC will be sent unordered.
+			// When the relevant users are migrated, we should remove the migration branch and enforce the presence of sender
+			// Removing it will allow the rest of the diagnostic code to emit the appropriate errors.
+			const bool bHasSenderAvailable = GSenderActor.IsSet();
 
-			if (GSenderStack.Num() == 0)
+			if (bIsUnordered)
 			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+				SenderInfo.Entity = WorkerEntityId;
 			}
-			else
+			else if (!bHasSenderAvailable)
 			{
-				SenderActor = GSenderStack.Last();
-			}
+				// Migration branch
+				if (bIsNetWriteFence)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error,
+						   TEXT("Net write fence will be dropped because no sender was provided. Function : %s, Target : %s"),
+						   *Function->GetName(), *Actor->GetName());
+					return;
+				}
+				else
+				{
+					UE_LOG(LogSpatialOSNetDriver, Warning,
+						   TEXT("Ordered reliable RPC will be sent unordered because no sender was provided. Function : %s, Target : %s"),
+						   *Function->GetName(), *Actor->GetName());
 
-			if (SenderActor == nullptr)
-			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor set to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+					SenderInfo.Entity = WorkerEntityId;
+				}
 			}
-
-			if (SenderActor && !SenderActor->HasAuthority())
+			else
 			{
-				UE_LOG(LogSpatialOSNetDriver, Error, TEXT("No authority on sender Actor %s to call RPC %s on target %s"),
-					   *SenderActor->GetName(), *Function->GetName(), *Actor->GetName());
-				return;
-			}
+				// Long term branch
+				AActor* SenderActor = nullptr;
+
+				if (!GSenderActor.IsSet())
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor. Function : %s, Target : %s"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+				else
+				{
+					SenderActorDesc& Desc = GSenderActor.GetValue();
+					SenderActor = Desc.Actor;
+
+					if (bIsOnlyNetWriteFence && Desc.Kind != SenderActorDesc::Dependent
+						|| (!bIsNetWriteFence && Desc.Kind == SenderActorDesc::Dependent))
+					{
+						UE_LOG(LogSpatialOSNetDriver, Error,
+							   TEXT("Wrong kind of sender Actor. Check that the right AActor function was used with the right kind of RPC "
+									"(CrossServer and NetWriteFence). Function : %s, Target : %s"),
+							   *Function->GetName(), *Actor->GetName());
+						return;
```

Comment:
Shouldn't we call GSenderActor.Reset(); before returning?

Reply:
Ehh, maybe it would not change much... I actually do not need to clean it at all because once in ProcessRPC there is no chance for confusion until the caller cleans it up itself.